components/region_cache_memory_engine/src/gc.rs

```diff
@@ -0,0 +1,524 @@
+        write_cf_handle: Arc<Skiplist<InternalKeyComparator, GlobalMemoryLimiter>>,
+    ) -> Self {
+        Self {
+            safe_point,
+            default_cf_handle,
+            write_cf_handle,
+            unique_key: 0,
+            filtered_write_key_size: 0,
+            filtered_write_key_buffer: Vec::with_capacity(100),
+            mvcc_key_prefix: vec![],
+            delete_versions: 0,
+            versions: 0,
+            filtered: 0,
+            cached_delete_key: None,
+            mvcc_rollback_and_locks: 0,
+            remove_older: false,
+        }
+    }
+
+    fn filter(&mut self, key: &[u8], value: &[u8]) -> Result<(), String> {
+        let InternalKey { user_key, .. } = decode_key(key);
+
+        let (mvcc_key_prefix, commit_ts) = split_ts(user_key)?;
+        if commit_ts > self.safe_point {
+            return Ok(());
+        }
+
+        self.versions += 1;
+        if self.mvcc_key_prefix != mvcc_key_prefix {
+            self.unique_key += 1;
+            self.mvcc_key_prefix.clear();
+            self.mvcc_key_prefix.extend_from_slice(mvcc_key_prefix);
+            self.remove_older = false;
+            self.remove_cached_delete_key();
+        }
+
+        let mut filtered = self.remove_older;
+        let write = parse_write(value)?;
+        if !self.remove_older {
+            match write.write_type {
+                WriteType::Rollback | WriteType::Lock => {
+                    self.mvcc_rollback_and_locks += 1;
+                    filtered = true;
+                }
+                WriteType::Put => self.remove_older = true,
+                WriteType::Delete => {
+                    self.delete_versions += 1;
+                    self.remove_older = true;
+
+                    // The first mvcc type below safe point is the mvcc delete. We should delay to
+                    // remove it until all the followings with the same user key have been deleted
+                    // to avoid older version apper.
+                    self.cached_delete_key = Some(key.to_vec());
+                }
+            }
+        }
+
+        if !filtered {
+            return Ok(());
+        }
+        self.filtered += 1;
+        // we cannot delete it directly as it's the node the iterator currently points
+        // to, so we cache it in the buffer and delete them later.
+        self.filter_write_cf_key(key);
+        self.handle_filtered_write(write)?;
+
+        Ok(())
+    }
+
+    fn remove_cached_delete_key(&mut self) {
+        if let Some(cached_delete_key) = self.cached_delete_key.take() {
+            // remove the older versions first
+            self.remove_filtered_key();
+            self.write_cf_handle.remove(&cached_delete_key);
+        }
+    }
+
+    fn filter_write_cf_key(&mut self, key: &[u8]) {
+        if self.filtered_write_key_size + key.len() >= EVICTION_KEY_BUFFER_LIMIT {
+            self.remove_filtered_key();
```

Comment:
As the key is deleted one by one, is there any benefits to cache it and delete later?

Reply:
No yet, each delete in skiplist reflects a entire search. I cache it here because we want to delete a key that the iterator is currently pointting. I may implement a delete range later in skiplist.

Reply:
> "I cache it here because we want to delete a key that the iterator is currently pointing"

This should be guaranteed by the GC is always deleting keys before the safe-point, which means no one is reading it. 
Otherwise, the current code  "if self.filtered_write_key_size + key.len() >= EVICTION_KEY_BUFFER_LIMIT" will break it anyway if it happens the `key` is used but the filtered_write_key_size exceeding EVICTION_KEY_BUFFER_LIMIT